### DISCUSSION TITLE
Berry MQTT Subscribe

### DISCUSSION THREAD

**Discussion by Axel830** [arendst/Tasmota#14856] [Q&A] [answered]
Hi,

I followed the example in #14403  and tried to subscribe to a mqtt topic with the subscribe command.
I also created a Berry driver to test:
`class mqttTest
  def mqtt_data(topic, idx, payload_s, payload_b)
    print(topic, idx, payload_s, payload_b)
    return(false)
  end
end
tasmota.add_driver(mqttTest())`

But the berry driver never gets called. 
After digging around in the sources and adding an additional log-statement it seems the rules engine consumes every subscribed topic and then the berry driver never gets called.
Tried with Rule1 enabled and disabled

Am I missing something?

**Comment by s-hadinger** [comment, marked as answer]
I'm currently adding a proper Berry MQTT subscription. Stay tuned

**Comment by boozeman** [comment]
Is there yet Any examples for berry mqtt.subscribe? Subscribe topic and get value as variable etc.

**Reply by sfromis** [reply]
Here's an example

```
class TestSub
  def init()
    import mqtt
    mqtt.subscribe('tele/another.tasmota.device/SENSOR')
    tasmota.add_driver(self)
  end
  def mqtt_data(topic, idx, strdata, bindata)
    import json
    var sensorvalues = json.load(strdata)
    print(sensorvalues['BME280']['Temperature'], sensorvalues['TCS34725']['Illuminance'])
    return true
  end
end
var testsub=TestSub()
```

**Reply by barbudor** [reply]
Don't you need to check that topic is your topic of interrest in `mqtt_data` ?
I think Berry is forwarding every MQTT message to every driver and it's up to the driver to handle or not the driver and return true/false accordingly.

**Reply by sfromis** [reply]
Depends, while it indeed would be a good idea in a broader perspective, like if there's further subscriptions, with only one subscription going on, i've now seen it run for a couple of hours without anything messing it up. Likewise, it would make sense to check if the incoming JSON contains the fields before accessing them from the map.

**Reply by sfromis** [reply]
Here's a more robust way of coding.....
```
class TestSub
  var topic
  def init()
    self.topic = 'tele/another.tasmota.device/SENSOR'
    import mqtt
    mqtt.subscribe(self.topic)
    tasmota.add_driver(self)
  end
  def mqtt_data(topic, idx, strdata, bindata)
    var hit = topic == self.topic
    if hit
      import json
      var sensorvalues = json.load(strdata)
      print(sensorvalues.find('BME280', {}).find('Temperature', 'missing temperature'))
    end
    return hit
  end
end
var testsub=TestSub()
```

**Reply by gonzzzales** [reply]
Hi all. I tryed to do this, but unsuccessful. 
`
class Nextion : Driver

def mqtt_data(topic, idx, strdata, bindata)
        print('1111');
        import json
        var sensorvalues = json.load(strdata)
        print(sensorvalues)
        print(sensorvalues['Temperature'])
        return true
    end
    
def init()
        log("NXP: Initializing Driver")
        self.ser = serial(17, 16, 115200, serial.SERIAL_8N1)
        self.sendnx('DRAKJHSUYDGBNCJHGJKSHBDN')
        self.sendnx('rest')
        self.flash_mode = 0
        import mqtt
        mqtt.subscribe('input/temp')
    end
end

var nextion = Nextion()

tasmota.add_driver(nextion)

`

So, i try to send information to topic input/temp, but dont see in console anything at all.
Please help to understand!!!

**Reply by sfromis** [reply]
You do not show where that `input/temp` topic comes from, how it is published....
FTR, subscribe is for getting publish data from *other* devices, not same device as where you do the subscribe.

**Reply by gonzzzales** [reply]
> You do not show where that `input/temp` topic comes from, how it is published.... FTR, subscribe is for getting publish data from _other_ devices, not same device as where you do the subscribe.
I try to publish information from MQTTfx program. I dont know why, but it work when my topic is "cmnd/tasmota_0C82EC/SET"

**Reply by sfromis** [reply]
At least with current Tasmota, your code cannot work. 
You assign to `self.ser` and `self.flash_mode` without having such instance vars created.
If you're not using current Tasmota, please upgrade before trying anything at all.

**Reply by gonzzzales** [reply]
Is it possible to know %topic% of device? Or to use some macros which returns "tasmota_0C82EC"

**Reply by sfromis** [reply]
`topic=tasmota.cmd('Topic')['Topic']`

**Reply by gonzzzales** [reply]
thank you. work great!!!

**Reply by gonzzzales** [reply]
> Here's a more robust way of coding.....
> 
> ```
> class TestSub
>   var topic
>   def init()
>     self.topic = 'tele/another.tasmota.device/SENSOR'
>     import mqtt
>     mqtt.subscribe(self.topic)
>     tasmota.add_driver(self)
>   end
>   def mqtt_data(topic, idx, strdata, bindata)
>     var hit = topic == self.topic
>     if hit
>       import json
>       var sensorvalues = json.load(strdata)
>       print(sensorvalues.find('BME280', {}).find('Temperature', 'missing temperature'))
>     end
>     return hit
>   end
> end
> var testsub=TestSub()
> ```

Is it possible to subscribe to custom topic, for example 'cluster/mqtt'?

**Reply by sfromis** [reply]
That example is already what you need for that. I just used an existing SENSOR topic for simplicity, and easier understanding of the JSON parsing.

However, by now I'd recommend using the updated simpler approach, no longer needing to be added as a driver, and no need to check if the topic is the right one.
```
class TestSub
  def init()
    var subtopic = 'tele/another.tasmota.device/SENSOR'
    import mqtt
    mqtt.subscribe(subtopic, self.process_data)
  end
  static def process_data(topic, idx, data, databytes)
    import json
    var sensorvalues = json.load(data)
    print(sensorvalues.find('BME280', {}).find('Temperature', 'missing temperature'))
    return true
  end
end
var testsub=TestSub()
```
Of course, instead of `process_data`, you can also do it all directly in a closure defined in the `mqtt.subscribe` call, for even shorter/simpler code. This also means that you need no class+instance for a basic setup to work.
```
var subtopic = 'tele/another.tasmota.device/SENSOR'
import mqtt
mqtt.subscribe(subtopic,
  def (topic,idx,data,databytes) 
    import json
    var sensorvalues = json.load(data)
    print(sensorvalues.find('BME280', {}).find('Temperature', 'missing temperature'))
    return true
  end)
```

**Reply by s-hadinger** [reply]
That's much better. I think you forgot the closure to capture `self`:

`mqtt.subscribe(topic, /-> self.process_data)`

**Reply by sfromis** [reply]
No, I did not forget that, instead process_data is a static method. To avoid snafus, I did test the code example before posting :-) 

Arguably, not having it as a static method may be preferable, to allow for instance variable access.

**Reply by gonzzzales** [reply]
Dont work for me((

`class TestSub
  def init()
    var subtopic = 'cluster/mqtt'
    import mqtt
    mqtt.subscribe(subtopic, self.process_data)
  end
  static def process_data(topic, idx, data, databytes)
    print(data)
    return true
  end
end
var testsub=TestSub()`

00:00:00.270 BRY: Berry initialized, RAM used=3715 bytes
00:00:00.300 Project tasmota - Tasmota Version 11.1.0.1(tasmota)-2_0_3(2022-04-17T21:26:37)
00:00:00.536 BRY: failed to run compiled code 'value_error' - Unexpected 'function'

**Reply by gonzzzales** [reply]
when i do this - mqtt.subscribe(subtopic) - i dont have any errors by reboot, but dont see any reaction for publishing in mqtt topic. When i do this - 
mqtt.subscribe(subtopic,
            def (topic,idx,data,databytes) 
            print(data)
            return true
        end)
i have error - failed to run compiled code 'value_error' - Unexpected 'function'

**Reply by gonzzzales** [reply]
@sfromis what is your tasmota version?

**Reply by sfromis** [reply]
When I referred to "by now" and "updated", that was meant as a hint to not just keep running the same binary, but update to current dev binary, which is what I run. The enhancement is from 6 days ago. https://github.com/arendst/Tasmota/pull/15487

**Reply by gonzzzales** [reply]
I have updated the firmware. Now it boot without errors, but still i dont see in console reaction for publishing string in mqtt topic
```
class TestSub
  def init()
    var subtopic = 'tele/mqtt/get'
    import mqtt
    mqtt.subscribe(subtopic,
      def (topic,idx,data,databytes)
          print(data)
          return true
    end)
  end
end
var testsub=TestSub() 
```
Am i right, this code print in concole the payload of subscribed topic?

**Reply by gonzzzales** [reply]
I cant understand, what is wrong((

this is exapmle from https://github.com/arendst/Tasmota/pull/15487, nothing in console. 

```
def f(topic, idx, payload_s, payload_b)
  print(">>>", payload_s)
  return true
end

def g(topic, idx, payload_s, payload_b)
  print("+++", payload_s)
  return true
end

import mqtt
mqtt.subscribe("tele/mqtt/get", f)
mqtt.subscribe("tele/mqtt/set", g)
```
I reach mqtt messages only, when i publish it in full_topic of tasmota

**Reply by sfromis** [reply]
> Am i right, this code print in concole the payload of subscribed topic?

Yes, I just did a test with that exact code snippet (only adjusting the topic string a bit), and it worked as expected when doing a publish command from another Tasmota device (or even the same one).

**Reply by gonzzzales** [reply]
What the device do you have? I working with Sonoff NSPanel

**Reply by sfromis** [reply]
However, when copying the code snippet to `autoexec.be`, it no longer worked, @s-hadinger 
(A print inserted into the `init()` method, after the `mqtt.subscribe`, confirmed that it was executed)
Maybe there is a problem with the Berry-sourced subscribe being executed before MQTT is connected?

Weirdly, I saw autoexec.bec getting updated with a wrong time stamp, in the future from when I saved
```
autoexec.bec 2022-05-06T12:00:11      306
autoexec.be  2022-05-06T11:42:18      239
```

**Reply by gonzzzales** [reply]
Firmware version 11.1.0.2

**Reply by gonzzzales** [reply]
> However, when copying the code snippet to `autoexec.be`, it no longer worked, @s-hadinger

yes, you are right, in berry scripting console it work, but not in autoexec.be. I always tryed to do it in autoexec.be

**Reply by barbudor** [reply]
> ```
> autoexec.bec 2022-05-06T12:00:11      306
> autoexec.be  2022-05-06T11:42:18      239
> ```

The Bec is not compiled on save but on load of the be, 
So it took you 18 minutes to restart your device after saving 🤔

**Reply by sfromis** [reply]
> So it took you 18 minutes to restart your device after saving

No.

Restart was right after me saving, and my post is timestamped 11:46, long before 12:00. Yes, edited after creation, but not that part :-) 

I suspect that the timestamp confusion may be related to `autoexec.be` being loaded/compiled before Wifi+NTP connection. Still a weird timestamp to get....

**Reply by gonzzzales** [reply]
Now waiting for @s-hadinger

**Reply by s-hadinger** [reply]
You can't subscribe to MQTT topics in `autoexec.be` because the code is ran before the MQTT connection is up.

You need to postpone the subscribes to an event indicating that MQTT is up.

**Reply by s-hadinger** [reply]
The trigger should be:

``` berry
def f(topic, idx, payload_s, payload_b)
  print(">>>", payload_s)
  return true
end

def g(topic, idx, payload_s, payload_b)
  print("+++", payload_s)
  return true
end

def subscribes()
  import mqtt
  mqtt.subscribe("tele/mqtt/get", f)
  mqtt.subscribe("tele/mqtt/set", g)
end

tasmota.add_rule("MQTT#Connected=1", subscribes)
```

Which makes me think that there might be a problem of dangling closures when MQTT reconnect occurs.

**Reply by sfromis** [reply]
```
tasmota.add_rule('Mqtt#Connected', 
  def () 
    import testsub 
    tasmota.remove_rule('Mqtt#Connected')
  end)
```

**Reply by s-hadinger** [reply]
What happens if the Mqtt broker restarts?

**Comment by MubiTec** [comment]
Does anybody has another working example for me which I can but in my function? I do not get it work and don't know why.
Maybe another example will help me.

**Reply by sfromis** [reply]
Due to improvements, you should not use the outdated info in this thread, but instead much simpler, like this minimal example:
```
import mqtt
mqtt.subscribe(subtopic, /topic, idx, payload, bindata -> print(topic, payload))
```
Longer form, nearly equivalent:
```
def process(topic, idx, payload, bindata)
  print(topic, payload)
end
import mqtt
mqtt.subscribe(subtopic, process)
```

In both cases, `idx` is not useful, and `bindata` only in case of you getting binary data via the subscription. Wildcard topics like `tele/+/SENSOR` should also work.

The older concerns about having to do subscription management, and resubscribe after reconnections no longer applies, as this is now handled by simply using `mqtt.subscribe` (including Wifi outages and such). You can even do the `mqtt.subscribe` in `autoexec.be` before MQTT is connected.

**Reply by MubiTec** [reply]
Ok, thank you.
I Tried it this way in my function which is startet one time per second.
But I do not see anything, neither in console nor in berry console :-(

import mqtt
mqtt.subscribe("tele/tasmota_A02CBC/SENSOR", /topic, idx, payload, bindata -> print(topic, payload))

**Reply by sfromis** [reply]
You should absolutely **not** do the *subscribe* once a second, only once a boot. And the code worked for me (obviously with a topic to match my setup). And messages only arrives as they are published by the other device, the default TelePeriod is 300 seconds.

**Reply by MubiTec** [reply]
Ok, now I put it in autoexec.be and do not see anything. Should it be in the console? 
Test.be is another Programm, which works properly 

load("test.be")

import mqtt
mqtt.subscribe("tele/tasmota_A02CBC/SENSOR", /topic, idx, payload, bindata -> print(topic, payload))

**Reply by sfromis** [reply]
With that exact code in `autoexec.be` (no other things loaded), I'm seeing messages both on the Berry REPL console, and the classic Tasmota console, in the latter I also get messages like this after restart, and again after MQTT connection established:
`MQT: Subscribe to tele/xxxxxxxx/SENSOR`

Of course, you should make sure to have the basic details right, like both devices being in session with the same MQTT broker, and that the other device actually is publishing messages to the specific topic; such activity is visible in the console there.

**Reply by MubiTec** [reply]
Hmm, strange. IT is the same Broker and it is publishing and I do not see the substribe to...
I have Version 12.3.1 installed from tasmota Page. Or do I need a special Version like sensors or something like that?

**Reply by sfromis** [reply]
No, the `mqtt` module for Berry should be available in all standard builds. I'm using 12.4, but 12.3.1 should be fine too. The console log at boot may be illustrative about other things happening, or maybe *not* happening as they should. You could also monitor MQTT activity in your broker by whatever MQTT tool you prefer to use.

One potential risk factor is if you've done "security theater" in your broker to limit what can work.

**Reply by MubiTec** [reply]
Hmm, I Update now to 12.4, and still do not see anything. IT is Displayed in console that autoexec is loaded successfully, but nothing happens regarding subscribe.
I dont See any
MQT: Subscribe to tele/xxxxxxxx/SENSOR

I use the 12.4 from:
http://ota.tasmota.com/tasmota32/release/tasmota32s2.bin

Has anybody anymore ideas, whats ging wrong at my configuration?

**Reply by sfromis** [reply]
Not knowing your configuration, hard to tell what's going wrong.

I'd try inserting something like a `print('mqtt.subscribe executed')` right after the subscribe, to verify execution. And for good measure, activate `Weblog 3` for details in log.
